cg/apps/demultiplex/sample_sheet/api.py

```diff
@@ -0,0 +1,166 @@
+                flow_cell_sample_type=flow_cell.sample_type,
+            )
+        )
+        if not lims_samples:
+            message: str = f"Could not find any samples in LIMS for {flow_cell.id}"
+            LOG.warning(message)
+            raise SampleSheetError(message)
+        return create_sample_sheet_content(flow_cell=flow_cell, lims_samples=lims_samples)
+
+    def create_sample_sheet_file(self, flow_cell: FlowCellDirectoryData) -> None:
+        """Create a valid sample sheet in the flow cell directory and add it to Housekeeper."""
+        sample_sheet_content: list[list[str]] = self.get_sample_sheet_content(flow_cell)
+        if not self.force:
+            self.validator.validate_sample_sheet_from_content(sample_sheet_content)
+        LOG.info(f"Writing sample sheet to {flow_cell.sample_sheet_path.resolve()}")
+        if self.dry_run:
+            click.echo(
+                WriteStream.write_stream_from_content(
+                    file_format=FileFormat.CSV, content=sample_sheet_content
+                )
+            )
+            return
+        WriteFile.write_file_from_content(
+            content=sample_sheet_content,
+            file_format=FileFormat.CSV,
+            file_path=flow_cell.sample_sheet_path,
+        )
+        add_and_include_sample_sheet_path_to_housekeeper(
+            flow_cell_directory=flow_cell.path, flow_cell_name=flow_cell.id, hk_api=self.hk_api
+        )
+
+    def validate(self, sample_sheet_path: Path) -> None:
+        """Validate a sample sheet."""
+        self.validator.validate_sample_sheet_from_file(sample_sheet_path)
+
+    def get_or_create_sample_sheet(self, flow_cell_name: str, bcl_converter: str) -> None:
+        """
+        Ensure that a valid sample sheet is present in the flow cell directory.
+        If a valid sample sheet for the flow cell is present in Housekeeper, the function hard-links
+        it to the flow cell directory. If not and if the flow cell directory has a valid sample
+        sheet, the function adds and includes it to Housekeeper. If neither is present, the
+        function creates a sample sheet for the flow cell and adds and includes it to Housekeeper.
+        """
+        flow_cell: FlowCellDirectoryData = self.get_flow_cell(
+            flow_cell_name=flow_cell_name, bcl_converter=bcl_converter
+        )
+        if hk_sample_sheet_path := self.get_valid_sample_sheet_path_from_hk(flow_cell.id):
+            LOG.info(
+                "Sample sheet already exists in Housekeeper. Hard-linking it to flow cell directory"
+            )
+            if not self.dry_run:
+                os.link(src=hk_sample_sheet_path, dst=flow_cell.sample_sheet_path)
+        elif self.get_valid_sample_sheet_path(flow_cell.sample_sheet_path):
+            LOG.info("Sample sheet already exists in flow cell directory. Adding to Housekeeper")
+            if not self.dry_run:
+                add_and_include_sample_sheet_path_to_housekeeper(
```

Comment:
Side track but not sure what the difference is between add and include and whether the caller should care about the difference

Reply:
`add` only adds the file to the database, `include` links the file to the housekeeper bundle directory. This terminology distinction is very common in the `housekeeper_storage_functions` module. I am not sure why would we want to just `add` something to housekeeper but the functions are still there 🤷‍♂️ . The renaming is something that requires more discussion and therefore out of the scope of this PR

Reply:
I don't think renaming things that have names leaking implementation details requires discussion. Up to you 👍

Reply:
In any way, it is out of the scope of this pr